File: backend/routes/cart_routes.py
"""Multi-vendor cart routes.

Single `carts` collection: one doc per user with items grouped by vendor. UX
is a single unified cart in the drawer; checkout splits per vendor into one
`deal` + a chat message per vendor group.
"""
from __future__ import annotations
from datetime import datetime, timezone
from bson import ObjectId
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from database import get_db
from middleware.auth_middleware import require_auth

router = APIRouter(prefix="/v1/cart", tags=["cart"])


class AddItemBody(BaseModel):
    listing_id: str = Field(min_length=8)
    quantity: int = Field(1, ge=1, le=99)
    variant_selection: dict | None = None


class UpdateItemBody(BaseModel):
    quantity: int = Field(ge=1, le=99)


async def _get_cart(user_id: str) -> dict:
    db = get_db()
    doc = await db.carts.find_one({"user_id": user_id})
    if not doc:
        now = datetime.now(timezone.utc).isoformat()
        doc = {"user_id": user_id, "items": [], "created_at": now, "updated_at": now}
        res = await db.carts.insert_one(doc)
        doc["_id"] = res.inserted_id
    return doc


async def _hydrate(cart: dict) -> dict:
    """Attach listing + vendor snapshots so the UI can render without extra calls."""
    db = get_db()
    ids = [ObjectId(i["listing_id"]) for i in cart.get("items", []) if ObjectId.is_valid(i.get("listing_id", ""))]
    lookup = {}
    if ids:
        async for l in db.listings.find({"_id": {"$in": ids}, "is_deleted": {"$ne": True}}):
            lookup[str(l["_id"])] = l
    groups: dict = {}
    total = 0
    for it in cart.get("items", []):
        li = lookup.get(it["listing_id"])
        if not li:
            continue
        vendor_id = str(li.get("vendor_id"))
        price = float(li.get("offer_price") or li.get("price") or 0)
        line = price * int(it.get("quantity", 1))
        total += line
        item_out = {
            "listing_id": it["listing_id"],
            "quantity": int(it.get("quantity", 1)),
            "variant_selection": it.get("variant_selection"),
            "title": li.get("title"),
            "slug": li.get("slug"),
            "price": price,
            "line_total": line,
            "image": (li.get("images") or [{}])[0].get("url"),
        }
        groups.setdefault(vendor_id, {"vendor_id": vendor_id, "items": [], "subtotal": 0.0})
        groups[vendor_id]["items"].append(item_out)
        groups[vendor_id]["subtotal"] += line

    if groups:
        vendor_ids = [ObjectId(v) for v in groups if ObjectId.is_valid(v)]
        async for u in db.users.find({"_id": {"$in": vendor_ids}}, {"name": 1, "profile_pic": 1}):
            groups[str(u["_id"])]["vendor"] = {
                "id": str(u["_id"]), "name": u.get("name"), "profile_pic": u.get("profile_pic"),
            }

    return {
        "id": str(cart["_id"]),
        "items": list(cart.get("items", [])),
        "groups": list(groups.values()),
        "total_items": sum(int(i.get("quantity", 1)) for i in cart.get("items", [])),
        "total_amount": total,
    }


@router.get("/me")
async def get_my_cart(user=Depends(require_auth)):
    cart = await _get_cart(str(user.id))
    return await _hydrate(cart)


@router.post("/me/add")
async def add_to_cart(body: AddItemBody, user=Depends(require_auth)):
    db = get_db()
    if not ObjectId.is_valid(body.listing_id):
        raise HTTPException(400, "Invalid listing id")
    li = await db.listings.find_one({"_id": ObjectId(body.listing_id), "is_deleted": {"$ne": True}})
    if not li:
        raise HTTPException(404, "Listing not found")
    now = datetime.now(timezone.utc).isoformat()
    cart = await _get_cart(str(user.id))
    items = cart.get("items", [])
    for it in items:
        if it["listing_id"] == body.listing_id:
            it["quantity"] = min(99, int(it.get("quantity", 1)) + body.quantity)
            it["variant_selection"] = body.variant_selection or it.get("variant_selection")
            break
    else:
        items.append({
            "listing_id": body.listing_id, "quantity": body.quantity,
            "variant_selection": body.variant_selection, "added_at": now,
        })
    await db.carts.update_one(
        {"_id": cart["_id"]},
        {"$set": {"items": items, "updated_at": now}},
    )
    cart["items"] = items
    return await _hydrate(cart)


@router.patch("/me/items/{listing_id}")
async def update_item(listing_id: str, body: UpdateItemBody, user=Depends(require_auth)):
    db = get_db()
    cart = await _get_cart(str(user.id))
    found = False
    for it in cart["items"]:
        if it["listing_id"] == listing_id:
            it["quantity"] = body.quantity
            found = True
    if not found:
        raise HTTPException(404, "Item not in cart")
    await db.carts.update_one({"_id": cart["_id"]},
        {"$set": {"items": cart["items"], "updated_at": datetime.now(timezone.utc).isoformat()}})
    return await _hydrate(cart)


@router.delete("/me/items/{listing_id}")
async def remove_item(listing_id: str, user=Depends(require_auth)):
    db = get_db()
    cart = await _get_cart(str(user.id))
    cart["items"] = [i for i in cart["items"] if i["listing_id"] != listing_id]
    await db.carts.update_one({"_id": cart["_id"]},
        {"$set": {"items": cart["items"], "updated_at": datetime.now(timezone.utc).isoformat()}})
    return await _hydrate(cart)


@router.post("/me/checkout")
async def checkout_cart(user=Depends(require_auth)):
    """Split cart per vendor → create one 'deal' per vendor + one chat thread + a
    summary message. Cart is cleared on success. Returns array of created deals.
    """
    db = get_db()
    cart = await _get_cart(str(user.id))
    hydrated = await _hydrate(cart)
    if not hydrated["groups"]:
        raise HTTPException(400, "Cart is empty")

    # Phase 7f (CHANGE 4): block checkout if ANY vendor in cart is unverified.
    # Vendor must have ≥1 approved KYC doc to accept order requests.
    from services import identity_service
    unverified_vendor_ids: list[str] = []
    unverified_vendor_names: list[str] = []
    for group in hydrated["groups"]:
        vid = group["vendor_id"]
        if not await identity_service.has_verified_identity(vid):
            unverified_vendor_ids.append(vid)
            unverified_vendor_names.append(group.get("vendor_name") or f"Vendor {vid[-4:]}")
    if unverified_vendor_ids:
        raise HTTPException(
            status_code=403,
            detail={
                "code": "vendor_unverified",
                "message": (
                    f"Cannot place this order — {len(unverified_vendor_ids)} vendor(s) "
                    "haven't verified their identity yet. Please ask them to verify to accept orders."
                ),
                "vendor_ids": unverified_vendor_ids,
                "vendor_names": unverified_vendor_names,
            },
        )

    from services import chat_service
    now = datetime.now(timezone.utc).isoformat()
    created = []
    for group in hydrated["groups"]:
        vendor_id = group["vendor_id"]
        subtotal = group["subtotal"]
        items_snapshot = [
            {"listing_id": i["listing_id"], "title": i["title"],
             "quantity": i["quantity"], "price": i["price"], "line_total": i["line_total"]}
            for i in group["items"]
        ]
        deal_doc = {
            "buyer_id": str(user.id), "vendor_id": vendor_id, "seller_id": vendor_id,
            "items": items_snapshot, "amount_paise": int(subtotal * 100),
            "status": "negotiating", "source": "cart_checkout",
            "created_at": now, "updated_at": now, "is_deleted": False,
        }
        res = await db.deals.insert_one(deal_doc)
        deal_id = str(res.inserted_id)

        # Chat thread + summary
        try:
            thread = await chat_service.get_or_create_thread(
                customer_id=str(user.id), vendor_id=vendor_id,
                listing_id=items_snapshot[0]["listing_id"],
            )
            summary_lines = [f"🛒 Order request — {len(items_snapshot)} item(s)"]
            for i in items_snapshot:
                summary_lines.append(f"  • {i['title']} × {i['quantity']} = ₹{i['line_total']:.0f}")
            summary_lines.append(f"Total: ₹{subtotal:,.0f}")
            await chat_service.send_message(
                thread_id=str(thread["_id"] if isinstance(thread, dict) else thread.id),
                sender_id=str(user.id),
                content="\n".join(summary_lines),
            )
        except Exception:  # noqa: BLE001
            pass

        created.append({"deal_id": deal_id, "vendor_id": vendor_id,
                        "amount_paise": deal_doc["amount_paise"],
                        "item_count": len(items_snapshot)})

    # Clear cart
    await db.carts.update_one({"_id": cart["_id"]}, {"$set": {"items": [], "updated_at": now}})
    return {"ok": True, "deals": created}

File: backend/src/routes/cart.routes.js
const express = require('express');
const mongoose = require('mongoose');
const { requireAuth } = require('../middleware/auth.middleware');
const Listing = require('../models/Listing');
const User = require('../models/User');
const Deal = require('../models/Deal');
const identityService = require('../services/identity.service');
const chatService = require('../services/chat.service');
const { catchAsync } = require('../utils/helpers');
const ApiError = require('../utils/ApiError');

const router = express.Router();

const getCartCollection = () => mongoose.connection.db.collection('carts');

const getCart = async (userId) => {
  const coll = getCartCollection();
  let doc = await coll.findOne({ user_id: userId });
  if (!doc) {
    const now = new Date().toISOString();
    doc = { user_id: userId, items: [], created_at: now, updated_at: now };
    const res = await coll.insertOne(doc);
    doc._id = res.insertedId;
  }
  return doc;
};

const hydrateCart = async (cart) => {
  const listingIds = (cart.items || [])
    .map(i => i.listing_id)
    .filter(id => mongoose.Types.ObjectId.isValid(id));

  const listings = listingIds.length > 0 ? await Listing.find({ _id: { $in: listingIds }, is_deleted: { $ne: true } }) : [];
  const lookup = {};
  for (const l of listings) {
    lookup[l._id.toString()] = l;
  }

  const missingIds = listingIds.filter(id => !lookup[id.toString()]);
  if (missingIds.length > 0) {
    const Reel = require('../models/Reel');
    const reels = await Reel.find({ _id: { $in: missingIds }, isDeleted: { $ne: true } }).populate('creator');
    for (const r of reels) {
      lookup[r._id.toString()] = {
        _id: r._id,
        vendor_id: r.creator?._id || r.creator,
        vendor: r.creator?._id || r.creator,
        title: r.caption || 'Reel Product',
        salePrice: Number(r.targetListing?.salePrice || r.salePrice || r.price || 0),
        price: Number(r.targetListing?.price || r.price || 0),
        images: [{ url: r.thumbnailUrl || (r.mediaUrls && r.mediaUrls[0]) || '' }],
      };
    }
  }

  const groups = {};
  let total = 0;

  for (const it of cart.items || []) {
    const li = lookup[it.listing_id];
    if (!li) continue;

    const vendorId = (li.vendor_id || li.vendor).toString();
    const priceCandidates = [
      li.sellingPrice,
      li.salePrice,
      li.offer_price,
      li.price,
      li.rate,
      li.cost,
      li.actualPrice,
      li.regularPrice,
      li.originalPrice,
      li.pricing?.amount,
      li.pricing?.price,
    ];
    const validPrice = priceCandidates.map(p => Number(p)).find(p => !isNaN(p) && p > 0);
    const price = validPrice || 0;
    const quantity = parseInt(it.quantity || 1, 10);
    const line = price * quantity;
    total += line;

    const itemOut = {
      listing_id: it.listing_id,
      quantity,
      variant_selection: it.variant_selection || null,
      title: li.title,
      slug: li.slug,
      price,
      line_total: line,
      image: li.images && li.images.length > 0 ? li.images[0].url : null,
    };

    if (!groups[vendorId]) {
      groups[vendorId] = { vendor_id: vendorId, items: [], subtotal: 0.0 };
    }
    groups[vendorId].items.push(itemOut);
    groups[vendorId].subtotal += line;
  }

  const vendorIds = Object.keys(groups);
  if (vendorIds.length > 0) {
    const users = await User.find({ _id: { $in: vendorIds } }).select('name profile_pic');
    for (const u of users) {
      const vid = u._id.toString();
      if (groups[vid]) {
        groups[vid].vendor = {
          id: vid,
          name: u.name,
          profile_pic: u.profile_pic || null,
        };
      }
    }
  }

  return {
    id: cart._id.toString(),
    items: cart.items || [],
    groups: Object.values(groups),
    total_items: (cart.items || []).reduce((sum, i) => sum + parseInt(i.quantity || 1, 10), 0),
    total_amount: total,
  };
};

router.get(['/', '/me'], requireAuth, catchAsync(async (req, res) => {
  const cart = await getCart(req.user._id.toString());
  const result = await hydrateCart(cart);
  res.json(result);
}));

router.post(['/add', '/me/add'], requireAuth, catchAsync(async (req, res) => {
  const listing_id = req.body.listing_id || req.body.listingId;
  const { quantity = 1, variant_selection } = req.body;
  if (!listing_id || !mongoose.Types.ObjectId.isValid(listing_id)) {
    throw ApiError.badRequest('Invalid listing id');
  }

  const li = await Listing.findOne({ _id: listing_id, is_deleted: { $ne: true } });
  if (!li) {
    const Reel = require('../models/Reel');
    const reel = await Reel.findOne({ _id: listing_id, isDeleted: { $ne: true } });
    if (!reel) {
      throw ApiError.notFound('Listing or product not found');
    }
  }

  const now = new Date().toISOString();
  const cart = await getCart(req.user._id.toString());
  const items = cart.items || [];
  let found = false;

  for (const it of items) {
    if (it.listing_id === listing_id) {
      it.quantity = Math.min(99, parseInt(it.quantity || 1, 10) + parseInt(quantity, 10));
      if (variant_selection) {
        it.variant_selection = variant_selection;
      }
      found = true;
      break;
    }
  }

  if (!found) {
    items.push({
      listing_id,
      quantity: parseInt(quantity, 10),
      variant_selection: variant_selection || null,
      added_at: now,
    });
  }

  const coll = getCartCollection();
  await coll.updateOne(
    { _id: cart._id },
    { $set: { items, updated_at: now } }
  );

  cart.items = items;
  const result = await hydrateCart(cart);
  res.json(result);
}));

router.patch(['/items/:listing_id', '/me/items/:listing_id'], requireAuth, catchAsync(async (req, res) => {
  const { listing_id } = req.params;
  const { quantity } = req.body;
  if (quantity === undefined || quantity < 1 || quantity > 99) {
    throw ApiError.badRequest('quantity must be between 1 and 99');
  }

  const cart = await getCart(req.user._id.toString());
  const items = cart.items || [];
  let found = false;

  for (const it of items) {
    if (it.listing_id === listing_id) {
      it.quantity = parseInt(quantity, 10);
      found = true;
      break;
    }
  }

  if (!found) {
    throw ApiError.notFound('Item not in cart');
  }

  const now = new Date().toISOString();
  const coll = getCartCollection();
  await coll.updateOne(
    { _id: cart._id },
    { $set: { items, updated_at: now } }
  );

  cart.items = items;
  const result = await hydrateCart(cart);
  res.json(result);
}));

router.delete(['/items/:listing_id', '/me/items/:listing_id'], requireAuth, catchAsync(async (req, res) => {
  const { listing_id } = req.params;
  const cart = await getCart(req.user._id.toString());
  const items = (cart.items || []).filter(i => i.listing_id !== listing_id);

  const now = new Date().toISOString();
  const coll = getCartCollection();
  await coll.updateOne(
    { _id: cart._id },
    { $set: { items, updated_at: now } }
  );

  cart.items = items;
  const result = await hydrateCart(cart);
  res.json(result);
}));

router.post(['/checkout', '/me/checkout'], requireAuth, catchAsync(async (req, res) => {
  const { couponCode, couponDiscount = 0, shippingCharges = 0, address, pincode } = req.body || {};
  const cart = await getCart(req.user._id.toString());
  const hydrated = await hydrateCart(cart);
  if (hydrated.groups.length === 0) {
    throw ApiError.badRequest('Cart is empty');
  }

  // Block checkout if any vendor is unverified
  const unverifiedVendorIds = [];
  const unverifiedVendorNames = [];
  for (const group of hydrated.groups) {
    const vid = group.vendor_id;
    const verified = await identityService.hasVerifiedIdentity(vid);
    if (!verified) {
      unverifiedVendorIds.push(vid);
      const vname = group.vendor ? group.vendor.name : `Vendor ${vid.slice(-4)}`;
      unverifiedVendorNames.push(vname);
    }
  }

  if (unverifiedVendorIds.length > 0) {
    return res.status(403).json({
      code: 'vendor_unverified',
      message: `Cannot place this order — ${unverifiedVendorIds.length} vendor(s) haven't verified their identity yet. Please ask them to verify to accept orders.`,
      vendor_ids: unverifiedVendorIds,
      vendor_names: unverifiedVendorNames,
    });
  }

  const now = new Date().toISOString();
  const created = [];
  const totalSubtotal = hydrated.total_amount || 0;
  const numGroups = hydrated.groups.length;

  for (let idx = 0; idx < hydrated.groups.length; idx++) {
    const group = hydrated.groups[idx];
    const vendorId = group.vendor_id;
    const subtotal = group.subtotal;
    const groupRatio = totalSubtotal > 0 ? subtotal / totalSubtotal : 1 / numGroups;
    const allocatedDiscount = Math.round((Number(couponDiscount) || 0) * groupRatio);
    const allocatedShipping = Math.round((Number(shippingCharges) || 0) * groupRatio);
    const groupFinalAmount = Math.max(0, subtotal - allocatedDiscount + allocatedShipping);

    const itemsSnapshot = group.items.map(i => ({
      listing_id: i.listing_id,
      title: i.title,
      quantity: i.quantity,
      price: i.price,
      line_total: i.line_total,
    }));

    const deal = await Deal.create({
      buyer_id: req.user._id.toString(),
      vendor_id: vendorId,
      seller_id: vendorId,
      items: itemsSnapshot,
      amount_paise: Math.round(groupFinalAmount * 100),
      item_total: subtotal,
      coupon_code: couponCode || null,
      coupon_discount: allocatedDiscount,
      shipping_charges: allocatedShipping,
      delivery_address: address || null,
      pincode: pincode || null,
      status: 'negotiating',
      source: 'cart_checkout',
      created_at: now,
      updated_at: now,
      is_deleted: false,
    });

    // Send summary chat message
    try {
      const thread = await chatService.getOrCreateThread(
        req.user._id.toString(),
        vendorId,
        itemsSnapshot[0].listing_id
      );
      const summaryLines = [`🛒 Order request — ${itemsSnapshot.length} item(s)`];
      for (const i of itemsSnapshot) {
        summaryLines.push(`  • ${i.title} x ${i.quantity} = ₹${Math.round(i.line_total)}`);
      }
      summaryLines.push(`Subtotal: ₹${Math.round(subtotal)}`);
      if (allocatedDiscount > 0) {
        summaryLines.push(`🏷️ Coupon Discount (${couponCode || 'PROMO'}): -₹${allocatedDiscount}`);
      }
      if (allocatedShipping > 0) {
        summaryLines.push(`🚚 Shipping Charges: ₹${allocatedShipping}`);
      } else {
        summaryLines.push(`🚚 Delivery: FREE`);
      }
      summaryLines.push(`Total Amount: ₹${Math.round(groupFinalAmount)}`);
      if (address) {
        summaryLines.push(`📍 Deliver to: ${address}`);
      }

      await chatService.sendMessage(
        thread._id.toString(),
        req.user._id.toString(),
        summaryLines.join('\n')
      );
    } catch (err) {
      // Non-fatal fallback
    }

    created.push({
      deal_id: deal._id.toString(),
      vendor_id: vendorId,
      amount_paise: deal.amount_paise,
      item_count: itemsSnapshot.length,
    });
  }

  // Clear cart
  const coll = getCartCollection();
  await coll.updateOne({ _id: cart._id }, { $set: { items: [], updated_at: now } });

  res.json({ ok: true, deals: created });
}));

module.exports = router;
